extern/storage-sealing/input.go

```diff
@@ -311,41 +311,45 @@ func (m *Sealing) SectorAddPieceToAny(ctx context.Context, size abi.UnpaddedPiec
 	}
 
 	m.inputLk.Lock()
-	if _, exist := m.pendingPieces[proposalCID(deal)]; exist {
+	if pp, exist := m.pendingPieces[proposalCID(deal)]; exist {
 		m.inputLk.Unlock()
-		return api.SectorOffset{}, xerrors.Errorf("piece for deal %s already pending", proposalCID(deal))
+		select {
+		case <-pp.doneCh:
+			res := pp.resp
+			return api.SectorOffset{Sector: res.sn, Offset: res.offset.Padded()}, res.err
+		case <-ctx.Done():
+			return api.SectorOffset{}, ctx.Err()
+		}
```

Comment:
Could put this into a helper method (e.g. `waitResp`) on `pendingPiece` for readability, and dedupe the same bit of code below.